### PR TITLE
[dua] PBBR should resolve DUA dest for packet initiated from its MTD child

### PR DESCRIPTION
**Pull request by simonlingoogle**
Supersedes #7120 to fix TH 5.9.21: DUA-TC-21

This commit fixes a bug that PBBR does not send `Address Query` and `Backbone Query` on behalf of MED Children, and add a test to make sure DUA forwarding works on PBBR:
- Make sure PBBR sends `Address Query` and `Backbone Query` on behalf of MED Child
- Make sure PBBR does not send `Address Query` when the destination is MED Child
- Make sure MED Children do not send `Address Query` by itself
- Verifiy DUA reachability between PBBR and MED Child
- Make sure PBBR reduce IPv6 TTL by 1 when forwarding DUA destined messages

### CODE REVIEW

File: src/core/thread/mesh_forwarder_ftd.cpp [hunk 1]
```diff
@@ -175,7 +182,36 @@ void MeshForwarder::HandleResolved(const Ip6::Address &aEid, Error aError)
...
+                 error = kErrorNoRoute);
+
+    IgnoreError(Get<AddressResolver>().Resolve(aDst, dstRloc16, /* aAllowAddressQuery */ false));
+    VerifyOrExit(dstRloc16 == Get<Mle::MleRouter>().GetRloc16(), error = kErrorNoRoute);
+
+    IgnoreError(aMessage.Read(Ip6::Header::kHopLimitFieldOffset, ttl));
+    ttl++;
+    aMessage.Write(Ip6::Header::kHopLimitFieldOffset, ttl);
+
+    IgnoreError(Get<Ip6::Ip6>().HandleDatagram(aMessage, nullptr, nullptr, false));
```
Comment (simonlingoogle) [inline src/core/thread/mesh_forwarder_ftd.cpp:209]:
```suggestion
    IgnoreError(Get<Ip6::Ip6>().HandleDatagram(aMessage, nullptr, nullptr, false));
```

The `Netif *aNetif` is current using `nullptr`, however this seems not totally correct:
- For messages sent by another device (e.g. MED child), `aNetif` should be set to `ThreadNetif`
- For messages sent by the PBBR itself, `aNetif` should be `nullptr`

The problem is that it's not easy to check if the message was initiated by PBBR itself or forwarded by PBBR at the current place because the information (receiving netif) was lost. 
Always using `nullptr` will bypass `mForwardingEnabled` checking, which maybe undesirable. 

Thoughts on how to address this issue? @librasungirl @abtink  @jwhui

Comment (simonlingoogle) [reply src/core/thread/mesh_forwarder_ftd.cpp:209]: BTW, we are also violating `mForwardingEnabled` check in Ip6::SendRaw:

```
    error = HandleDatagram(aMessage, nullptr, nullptr, true);
```

File: tests/scripts/thread-cert/backbone/test_dua_routing_med.py [hunk 2]
```diff
@@ -0,0 +1,278 @@
...
+            .must_next()
+        # ROUTER1 sends Address Answer for ROUTER1_DUA
+        pkts.filter_wpan_src64(ROUTER1).filter_coap_request(ADDR_NTF_URI, confirmable=True) \
+            .filter('thread_address.tlv.target_eid == {eid}', eid=ROUTER1_DUA) \
+            .must_next()
+        # PBBR forwards Ping Request to ROUTER1, decreasing TTL by 1
+        ping_request_pkts.filter_wpan_src64(PBBR).filter_ping_request(identifier=ping_id).must_next().must_verify(
+            'ipv6.hlim == {hlim}', hlim=ping_pkt.ipv6.hlim - 1)
+
+        # MED1 pings MED1
```
Comment (librasungirl) [inline tests/scripts/thread-cert/backbone/test_dua_routing_med.py:250]:
```suggestion
        # MED1 pings MED2
```

File: src/core/thread/mesh_forwarder_ftd.cpp (hunk 1 above)
Comment (librasungirl) [reply src/core/thread/mesh_forwarder_ftd.cpp:209]: For checking whether or not it's from PBBR itself, may check whether or not the source of the message belongs to PBBR?

Here if the packet dest to DUA in another network, `forwardThread` should be false, there might be no chance to bypass `mForwardingEnabled` checking.
https://github.com/openthread/openthread/blob/ca450e541636692909183773beba8f55591a7185/src/core/net/ip6.cpp#L1256-L1262

In my understanding, `mForwardingEnabled` should be specific for forwarding rule in mesh and only router is allowed to forward packet it received via thread radio in mesh. I think it should be fine for SendRaw, from a different link into mesh.

Comment (librasungirl) [reply src/core/thread/mesh_forwarder_ftd.cpp:209]: May check whether or not the source of the message belongs to PBBR to decide whether it's initiated by the PBBR itself or its MTD child?

Here if the packet destines to DUA in another network, `forwardThread` should be false, there might be no chance to bypass `mForwardingEnabled` checking.
https://github.com/openthread/openthread/blob/ca450e541636692909183773beba8f55591a7185/src/core/net/ip6.cpp#L1256-L1262

In my understanding, `mForwardingEnabled` should be specific for forwarding rule in mesh and only router is allowed to forward packet that it received via thread radio in mesh. I think it should be fine for `SendRaw`, from a different link into mesh.

**Comment by simonlingoogle** [comment]
@librasungirl 

> In my understanding, mForwardingEnabled should be specific for forwarding rule in mesh and only router is allowed to forward packet it received via thread radio in mesh. I think it should be fine for SendRaw, from a different link into mesh.

+1 

> Here if the packet destines to DUA in another network, forwardThread should be false, there might be no chance to bypass mForwardingEnabled checking.
+1 

> May check whether or not the source of the message belongs to PBBR to decide whether it's initiated by the PBBR itself or its MTD child?

Setting `netif` seems useless according to your description above. So, maybe just fine to use `nullptr` for simplicity.

File: src/core/thread/mesh_forwarder_ftd.cpp (hunk 1 above)
Comment (simonlingoogle) [reply src/core/thread/mesh_forwarder_ftd.cpp:209]: @librasungirl 

> In my understanding, mForwardingEnabled should be specific for forwarding rule in mesh and only router is allowed to forward packet it received via thread radio in mesh. I think it should be fine for SendRaw, from a different link into mesh.

+1 

> Here if the packet destines to DUA in another network, forwardThread should be false, there might be no chance to bypass mForwardingEnabled checking.

+1 

> May check whether or not the source of the message belongs to PBBR to decide whether it's initiated by the PBBR itself or its MTD child?

Setting `netif` seems useless according to your description above. So, maybe just fine to use `nullptr` for simplicity.

File: src/core/backbone_router/bbr_manager.cpp [hunk 3]
```diff
@@ -514,12 +514,14 @@ bool Manager::ShouldForwardDuaToBackbone(const Ip6::Address &aAddress)
     VerifyOrExit(Get<Local>().IsPrimary());
     VerifyOrExit(Get<Leader>().IsDomainUnicast(aAddress));
 
+    // Do not forward to Backbone if the DUA is registered on PBBR
     VerifyOrExit(!mNdProxyTable.IsRegistered(aAddress.GetIid()));
-
+    // Do not forward to Backbone if the DUA belongs to a MED Child (which may failed in DUA registration)
```
Comment (librasungirl) [inline src/core/backbone_router/bbr_manager.cpp:519]:
```suggestion
    // Do not forward to Backbone if the DUA belongs to a MTD Child (which may failed in DUA registration)
```

File: src/core/thread/mesh_forwarder_ftd.cpp (hunk 1 above)
Comment (librasungirl) [reply src/core/thread/mesh_forwarder_ftd.cpp:209]: no preference, either is fine to me.

File: src/core/backbone_router/bbr_manager.cpp (hunk 3 above)
Comment (librasungirl) [reply src/core/backbone_router/bbr_manager.cpp:519]: may consider add some test on this corner scenario in the future.

File: src/core/thread/mesh_forwarder_ftd.cpp [hunk 4]
```diff
@@ -160,6 +160,13 @@ void MeshForwarder::HandleResolved(const Ip6::Address &aEid, Error aError)
 
             if (aError == kErrorNone)
             {
+#if OPENTHREAD_CONFIG_BACKBONE_ROUTER_ENABLE
+                // Pass back to IPv6 layer for DUA destination resolved by Backbone Query
+                if (ForwardDuaToBackboneLink(*cur, ip6Dst) == kErrorNone)
+                {
+                    ExitNow();
```
Comment (jwhui) [inline src/core/thread/mesh_forwarder_ftd.cpp:167]: What if there are multiple packets enqueued for the same destination?

File: src/core/backbone_router/bbr_manager.cpp [hunk 5]
```diff
@@ -514,12 +514,14 @@ bool Manager::ShouldForwardDuaToBackbone(const Ip6::Address &aAddress)
     VerifyOrExit(Get<Local>().IsPrimary());
     VerifyOrExit(Get<Leader>().IsDomainUnicast(aAddress));
 
+    // Do not forward to Backbone if the DUA is registered on PBBR
     VerifyOrExit(!mNdProxyTable.IsRegistered(aAddress.GetIid()));
-
+    // Do not forward to Backbone if the DUA belongs to a MTD Child (which may failed in DUA registration)
```
Comment (jwhui) [inline src/core/backbone_router/bbr_manager.cpp:519]:
```suggestion
    // Do not forward to Backbone if the DUA belongs to a MTD Child (which may have failed in DUA registration)
```

File: src/core/thread/mesh_forwarder_ftd.cpp [hunk 6]
```diff
@@ -175,7 +182,36 @@ void MeshForwarder::HandleResolved(const Ip6::Address &aEid, Error aError)
...
+    uint8_t  ttl;
+
+    VerifyOrExit(Get<BackboneRouter::Local>().IsPrimary() && Get<BackboneRouter::Leader>().IsDomainUnicast(aDst),
+                 error = kErrorNoRoute);
+
+    IgnoreError(Get<AddressResolver>().Resolve(aDst, dstRloc16, /* aAllowAddressQuery */ false));
+    VerifyOrExit(dstRloc16 == Get<Mle::MleRouter>().GetRloc16(), error = kErrorNoRoute);
+
+    IgnoreError(aMessage.Read(Ip6::Header::kHopLimitFieldOffset, ttl));
+    ttl++;
```
Comment (jwhui) [inline src/core/thread/mesh_forwarder_ftd.cpp:206]: Add comment that this is to avoid decrementing twice?

File: src/core/thread/mesh_forwarder_ftd.cpp (hunk 4 above)
Comment (simonlingoogle) [reply src/core/thread/mesh_forwarder_ftd.cpp]: fixed

**Review by librasungirl — APPROVED** [review]
👍

**Review by suveshpratapa — COMMENTED** [review]
@simonlingoogle @librasungirl Reviewed it, looks good, but I have one question.

File: src/core/backbone_router/bbr_manager.cpp [hunk 7]
```diff
@@ -515,11 +515,11 @@ bool Manager::ShouldForwardDuaToBackbone(const Ip6::Address &aAddress)
     VerifyOrExit(Get<Leader>().IsDomainUnicast(aAddress));
 
     VerifyOrExit(!mNdProxyTable.IsRegistered(aAddress.GetIid()));
+    VerifyOrExit(Get<NeighborTable>().FindNeighbor(aAddress) == nullptr);
 
     error = Get<AddressResolver>().Resolve(aAddress, rloc16, /* aAllowAddressQuery */ false);
-    VerifyOrExit(error != kErrorNone || rloc16 == Get<Mle::MleRouter>().GetRloc16());
+    VerifyOrExit(error == kErrorNone && rloc16 == Get<Mle::MleRouter>().GetRloc16());
```
[the remark this answers is not in the document]
Comment (simonlingoogle) [reply src/core/backbone_router/bbr_manager.cpp:523]: If we get `kErrorNotFound`, we will exit with `forwardToBackbone` being false, which is what we want. 

`kErrorNotFound` means the Address Resolver does not have a cache entry for the DUA address, in such case, we should send Address Query and Backbone Query for the DUA. Note that by returning `forwardToBackbone=false`, the message will be re-forwared by the Mesh Forwarder, which will send Address Query and Backbone Query to resolve the DUA.

[the remark this answers is not in the document]
Comment (suveshpratapa) [reply src/core/backbone_router/bbr_manager.cpp:523]: Makes sense! Thank you for explaining the last part to me, that was unclear.

**Review by jwhui — APPROVED** [review]
